language-plutus-core/src/Language/PlutusCore/Constant/DynamicType.hs

```diff
@@ -0,0 +1,89 @@
+-- | Dynamic built-in types.
+
+{-# LANGUAGE DefaultSignatures #-}
+
+module Language.PlutusCore.Constant.DynamicType
+    ( KnownDynamicBuiltinType (..)
+    , PrettyDynamic (..)
+    ) where
+
+import           Language.PlutusCore.Lexer.Type
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Type
+import           PlutusPrelude
+
+import qualified Data.ByteString.Lazy.Char8     as BSL
+
+{- Note [Semantics of dynamic built-in types]
+We only allow dynamic built-in types that
+
+1. can be represented using static types in PLC. For example Haskell's 'Char' can be represented as
+@integer 4@ in PLC. This restriction makes the dynamic built-in types machinery somewhat similar to
+the @newtype@ machinery in Haskell. The reason for this restriction is that storing values of arbitrary
+types of a host language in the AST of a target language is commonly far from being trivial, hence we
+do not support this right now, but we plan to figure out a way to allow such extensions to the AST
+2. are of kind @*@. Dynamic built-in types that are not of kind @*@ can be encoded via recursive
+instances. For example:
+
+    instance KnownDynamicBuiltinType dyn => KnownDynamicBuiltinType [dyn] where
+        ...
+
+This is due to the fact that we use Haskell classes to assign semantics to dynamic built-in types and
+since it's anyway impossible to assign a meaning to an open PLC type, because you'd have to somehow
+interpret free variables, we're only interested in closed PLC types and those can be handled by
+recursive instances as shown above.
+
+Instance resolution is driven by Haskell types rather than names of PLC types.
+An alternative design would be something like
```

Comment:
I'd appreciate a bit more discussion of the *actual* design before contrasting it with alternatives!

Reply:
I've added a few lines, but I'm not sure what to describe there. Here are the lines:

```
An @KnownDynamicBuiltinType dyn@ instance provides

1. a way to encode @dyn@ as a PLC type ('getTypeEncoding')
2. a function that encodes values of type @dyn@ as PLC terms ('makeDynamicBuiltin')
3. a function that decodes PLC terms back to Haskell values ('readDynamicBuiltin')

The last two are ought to constitute an isomorphism (modulo 'Quote' and 'Maybe').
```

Is there anything else that needs to be described? I mean, this stuff is not complicated.